### ISSUE TITLE
segfault on ubuntu 16.04

### ISSUE THREAD

**Issue by sbabic** [lefcha/imapfilter#131]
Hi,
I have recently updated my server to 16.04.01 and imapfilter was upgraded from 2.5.2 to 2.6.4. It crashes. I have then compiled from TOT (v2.6.7), it crashes, too. I have starting bisecting and I have found that all releases starting with 2.5.7 are crashing when the program is invoked, and adding -v it happens after STARTTLS. Reverting the following commit let 2.5.7 working on Ubuntu 16.04:

commit 31fb50dab87662b65aaa8b16e3a1839300ea5d63
Author: Lefteris Chatzimparmpas lefcha@fastmail.net
Date:   Sun Nov 9 01:24:03 2014 +0200

```
Add support to automatically validate SSL connections' certificates
```

I stop searching, but it looks some sort of incompatibility with SSL libraries supplied with Ubuntu.

**Comment by Papafox** [comment]
I'm assuming that 16,04.01 refers to Ubuntu Xenial.  It has upgraded to
OpenSSL 1.0.2g-fips, which no longer has ssl2, ssl3 or tlsv1.0 support
compiled in.

The failure is at imapfilter.c:138.

137: #ifndef OPENSSL_NO_SSL3_METHOD
138:         SSL_CTX_load_verify_locations(ssl3ctx, cafile, capath);
139: #endif
140:         SSL_CTX_load_verify_locations(ssl23ctx, cafile, capath);
141:         SSL_CTX_load_verify_locations(tls1ctx, cafile, capath);
142: #if OPENSSL_VERSION_NUMBER >= 0x01000100fL
143:         SSL_CTX_load_verify_locations(tls11ctx, cafile, capath);
144:         SSL_CTX_load_verify_locations(tls12ctx, cafile, capath);
145: #endif

The value of ssl3ctx is null because as noted above, ssl3 is not available
in this version of OpenSSL and thus causes a segfault.

Matthew








————————————————————————————————————————————————————
*Matthew Donald*
m: +61-(0)431 607 492
e:  matthew@wsadmin.org

**Comment by lefcha** [comment]
But why the segfault? Shouldn't imapfilter exclude compiling in those lines of code, if the SSL3 is not built into the OpenSSL library? I'll look into this...

**Comment by justinabrahms** [comment]
I'm seeing this as well, if you need someone to help reproduce w/ test cases or similar.

**Comment by lefcha** [comment]
Is this still a problem with the latest version 2.6.9 released earlier this week?

**Comment by justinabrahms** [comment]
I'm still seeing it, yes.
![image](https://cloud.githubusercontent.com/assets/3853/21064841/131659ce-be12-11e6-8f27-50f1c276989b.png)

**Comment by justinabrahms** [comment]
I've reached the same conclusion as sbabic. v2.5 is a good tag. master is bad. The bisect script I had was this:


```
#!/bin/bash -x
LD_CONFIG=/usr/lib/x86_64-linux-gnu:$LD_CONFIG CPATH=/usr/include/lua5.3/:$INCLUDE_DIR make LIBLUA=-llua5.3
(src/imapfilter -e "options.timeout = 60" && exit 0) || exit 5
```

The final output of the bisect was:

```
31fb50dab87662b65aaa8b16e3a1839300ea5d63 is the first bad commit
commit 31fb50dab87662b65aaa8b16e3a1839300ea5d63
Author: Lefteris Chatzimparmpas <lefcha@fastmail.net>
Date:   Sun Nov 9 01:24:03 2014 +0200

    Add support to automatically validate SSL connections' certificates
    
    This finally adds support to validate SSL connections using trusted CA
    certificates installed in the system, usually in /etc/ssl/certs.  So the
    user doesn't have to manually accept and store certificates for most
    IMAP servers.
```

If you need me to repro or do any specific testing, I'm happy to help, but I don't know much about C++ / openssl, so I'd need some guidance.

**Comment by lefcha** [comment]
Can you check with the latest version; I just did a commit that should take care of such segfaults.

**Comment by justinabrahms** [comment]
I no longer get a segfault. :)

**Comment by lefcha** [comment]
Closed by 0a1a255fe768a25071f1b868f25d32bf20403956.
